## Exercises left for week 01 PM Prod
This exercises were left as a way to let students experiment with algortithms and see how capable they are at solving these. These exercises are
1. Knapsack Problem
2. Perfect Numbers
3. Binary Tree
4. Variance problem

### Knapsack problem
The knapsack problem is a problem  is a problem in [combinatorial optimization](https://en.wikipedia.org/wiki/Combinatorial_optimization): Given a set of items, each with a weight and a value, determine the number of each item to include in a collection so that the total weight is less than or equal to a given limit and the total value is as large as possible.

This is an interesting problem since is an [NP](https://www.youtube.com/watch?v=UR2oDYZ-Sao) pronblem, this means that there is not a polinomial way to resolve this porblem and it is hard. In simple words it means that there is not one alorithm that will always lead you to the right answer.

Still the problem is simple, if you have a knapsack with a maximum weiht, and a bunch of articles with weihts and values, how do I optimize the way I put them all together in mu knapsack in order for them to maximize my value.

Bellow I present an answer using recursion proramming and dynamic programming

**Disclosure: this was taken from one of my students Luis Manuel Martinez**

In [2]:
import csv
from collections import deque

In [20]:
%%time
## Recursion method
objects = deque()
max_w = 0
small = True
if small:
    data = 'knapsack.csv'
else:
    data = 'knapsack-2.csv'
        
def get_data():
    global max_w
    with open(data,'r') as f:
        csv_reader = csv.reader(f)
        i = 0
        for row in csv_reader:
            if i==0:
                i+=1
                continue
            if i == 1:
                i+=1
                max_w=int(row[2])
            objects.append((int(row[0]),int(row[1])))

def solve(w_left,i=0,v=0, knapsack = list()):
    if not w_left or i>=len(objects):
        return (v,knapsack)
    if(w_left>=objects[i][0]):
        return max(solve(w_left,i+1,v,knapsack),         solve(w_left-objects[i][0],        i+1,        v+objects[i][1],        list(knapsack+[(objects[i][0],objects[i][1])])        ))
    return solve(w_left,i+1,v,knapsack)


get_data()
k = (solve(w_left=max_w))

CPU times: user 53.7 ms, sys: 2.72 ms, total: 56.4 ms
Wall time: 57.4 ms


In [7]:
k

1458

As you can see this takes some time to answer, now let's take Dynamic prorammingg and compare.

In [29]:
def knapSack(W, wt, val, n):
    import numpy as np
    '''
    Input:
    W is the maximum weight
    wt is an array with the weights of the items
    val is an array of the values of the items
    n is the amount of items
    '''
    # First to make dynamic programming create a matrix with all possible combinations of n items with W weights 
    # basically what we are doping here is to predefine the matrix so we save the memory 
    # and time instead of doing recursive programming
    print("I am creating a %5d by %5d matrix to save the information" % (W+1, n+1))
    K = [[0] * (W + 1) for x in range(n + 1)] # Very pythonic way of writing this
    
 
    # Build table K[][] in bottom up manner
    for i in range(n + 1):
        for w in range(W + 1):
            if i == 0 or w == 0:
                K[i][w] = 0
            elif wt[i-1] <= w:
                K[i][w] = max(val[i-1]
                          + K[i-1][w-wt[i-1]],  
                              K[i-1][w])
            else:
                K[i][w] = K[i-1][w]
 
    return (np.array(K),K[n][W])

In [30]:
%%time
wt = [x[0] for x in objects]
val = [x[1] for x in objects]
n = len(val)
k = (knapSack(max_w, wt, val, n))

I am creating a   751 by    16 matrix to save the information
CPU times: user 9.27 ms, sys: 716 µs, total: 9.99 ms
Wall time: 9.52 ms


In [33]:
val = [60, 100, 120]
wt = [10, 20, 30]
W = 52
n = len(val)
print(knapSack(W, wt, val, n)[0])


I am creating a    53 by     4 matrix to save the information
[[  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0  60  60  60  60  60  60  60  60
   60  60  60  60  60  60  60  60  60  60  60  60  60  60  60  60  60  60
   60  60  60  60  60  60  60  60  60  60  60  60  60  60  60  60  60]
 [  0   0   0   0   0   0   0   0   0   0  60  60  60  60  60  60  60  60
   60  60 100 100 100 100 100 100 100 100 100 100 160 160 160 160 160 160
  160 160 160 160 160 160 160 160 160 160 160 160 160 160 160 160 160]
 [  0   0   0   0   0   0   0   0   0   0  60  60  60  60  60  60  60  60
   60  60 100 100 100 100 100 100 100 100 100 100 160 160 160 160 160 160
  160 160 160 160 180 180 180 180 180 180 180 180 180 180 220 220 220]]


As you can see with low amount of values the dynamic programming is faster, but with huge maximum capacity for the knapsack it will get slower and slower.

Still, for this las example we can see how the matrix works, array 0 checks for all the possible weight with the 0th elements, this is mainly done so it does not get an error when checking for the knapsack's n-1 element.

For the second array it checks the knapsack in all possible ways for the first element, the one with weight 10, that is why in the 10th element of the array is when we start seeing that the knapsack will have value 60. In the third row of the matrixz or the third array we see how with weight 10 it will carry the first element, at weight 20 it will only carry the second element and for the weight 30 it will carry both the elements. Following the same logic, Can you see what is happening in the fourth array and why the last 3 elements are the ones with value 220?

### Perfect Numbers
A [perfect number](https://en.wikipedia.org/wiki/Perfect_number) is such that the sum of all its divisors (without itself) gives the number at hand. For example 6 is a perfect number since 3+2+1 = 6. Now make a funcion that, given a number N prints all the perfect numbers less or equal to N that are perfect numbers. For this we will use [the fundamental theorem of arithmetic](https://brilliant.org/wiki/fundamental-theorem-of-arithmetic/#:~:text=The%20fundamental%20theorem%20of%20arithmetic,unique%20combination%20of%20prime%20numbers.) and [the theorem of composite numbers](https://proofwiki.org/wiki/Composite_Number_has_Prime_Factor_not_Greater_Than_its_Square_Root)

In [58]:
# Python program to print prime factors 

import math 

# A function to print all prime factors of 
# a given number n 
class PerfectNums:
    def __init__(self, N):
        self.n = N
        self.perfect_numbers = [] # this should be a queue
        
    def sumFactors(self, n): 
        sum_factor = 1 # this should be a queue
        
        i = 2
        # composite theorem
        while i <= math.sqrt(n): 
            if (n % i == 0) : 
                # If divisors are equal, print only one 
                if (n / i == i) : 
                    sum_factor += i
                else : 
                    # Otherwise print both 
                    sum_factor += i  
                    sum_factor += n/i 
            i = i + 1
        return sum_factor
    
    
        
    def GetPerfectNumbers(self):
        # we know that 6 is the first perfect number, so we need 
        for i in range(6,self.n+1):
            factors = self.sumFactors(i)
            
            if factors == i:
                self.perfect_numbers.append(i)
        return self.perfect_numbers

In [61]:
# Driver Program to test above function 
n = 10000
calcs = PerfectNums(n) 
calcs.GetPerfectNumbers()

[6, 28, 496, 8128]

In [3]:
class BinaryTree():

    def __init__(self,rootid):
      self.left = None
      self.right = None
      self.rootid = rootid

    def getLeftChild(self):
        return self.left
    def getRightChild(self):
        return self.right
    def setNodeValue(self,value):
        self.rootid = value
    def getNodeValue(self):
        return self.rootid

    def insertRight(self,newNode):
        if self.right == None:
            self.right = BinaryTree(newNode)
        else:
            tree = BinaryTree(newNode)
            tree.right = self.right
            self.right = tree

    def insertLeft(self,newNode):
        if self.left == None:
            self.left = BinaryTree(newNode)
        else:
            tree = BinaryTree(newNode)
            tree.left = self.left
            self.left = tree


def printTree(tree):
        if tree != None:
            printTree(tree.getLeftChild())
            print(tree.getNodeValue())
            printTree(tree.getRightChild())




myTree = BinaryTree(1000)
myTree.insertLeft(1)
myTree.insertRight(2)
myTree.insertRight(3)
printTree(myTree)

1
1000
3
2


### Get the variance of the event of getting two heads in a row.


### Answer
Saying that experiment A has probability p of happening (in this case getting heads) and experiment B has $q=1-p$ probability of happening, then:

If you wanted to calculate the probabilities, assuming $p$ is the probability of success and $q=1-p$, then the important thing is to notice is that for the experiment to work, then you need to have failed $x-2$ times the experiment and then have two consecutive experiments in the last two cases (similar to a geometric distribution). 

In this case, the probability of failing up to $x-2$ can be seen as all the combinations of experiments where the sum of the exponents of $p$ and $q$ sumed is $x-2$ but the exponent in $p$ has to be smaller than that of $q$ (this is due to the fact that if experiment A happens more times than experiemtn B, then at some moment in the sequence experiment A had to have happened two times in a row). Another constriction is that experiment A cannot happen last in the x-2 first experiments (if this was true, then the experiment would end after x-1 instead of at x). 

We need to ensure to not count the cases where p appears two consecutive times (or more) in the sequence. we accomplish this by getting ${x-2-k \choose k}$ (this can be proofed by deduction. Done bellow, but is not good and I believe it can be written better.)

[One could also use the probability generating function](https://math.stackexchange.com/questions/2845196/expected-number-of-coin-flips-for-2-consecutive-heads-for-first-time)

Then this means that 

$$P(X=x) = \bigg( \sum^{\frac{x-2}{2}}_{k=0}{x-2-k \choose k} q^{x-2-k}*p^{k}\bigg) * p^2$$

proof by deduction:

suppose we want the combinations of cases where the events with probability $p$ and $q$ happen $n$ times in total, but with the restriction of p never appearing in consecutive cases and it cannot be in the last position, then the easiest way to think about this is to imagine two events that you are drawing from $q$ and $pq$, then we can see what happens in each case to form our case (it is not necessary, but is easier to understand if put on paper)

* i=0
In this case we have n q events happening and 0 pq, then we have n choose 0 
$$f(n,p,q) = q*q*...*q = {n \choose 0} q^n$$
* i=1
For this case we are looking for the case where we hace n-2 q events and 1 pq event happening, then in total we have n-1 events happening. This means we have n-1 choose 1.
$$f(n,p,q) = p*q*...*q + q*p*...*q*q + ... + q*q*...*p*q = {n-1 \choose 1}*q^{n-1}*p^1$$
the reason why this is choose 1 in $n-1$ is because there are n-1 possible positions where 1 $p$ can be.
* i=2
For this case we are looking for the case where we hace n-4 q events and 2 pq event happening, then in total we have n-2 events happening. This means we have n-2 choose 2.
$$f(n,p,q) = {n-2 \choose 2}*q^{n-4}*(qp)^2 = {n-2 \choose 2}*q^{n-2}*p^2$$

getting the same logic as in the last case for the kth case we can say that
* i=k
$$f(n,p,q) = {n-k \choose k}*q^{n-k}*p^k$$ (where $k < n/2$)

(If you wanted to do a similar prove for the general case where the experiment ends after l events happening in the row, the same idea can be done as here, but with the events q, pq...q, p..pq and so on.)



In [15]:
def calculate_combinations(n, r):
    from math import factorial
    return factorial(n) // factorial(r) // factorial(n-r)
def calculate_probability(x, p=1/2):
    from math import floor
    cumulative = 0
    if x>1:
        for i in range(floor((x-2)/2)+1):
            cumulative += calculate_combinations(x-2-i,i)*(1-p)**(x-2-i)*p**(i)
        return cumulative*p**2
    else:
        return 0

In [16]:
calculate_probability(2)

0.25

In [17]:
def simulate_experiment(sims, p=1/2):
    trials = []
    for i in range(sims):
        counter = 0
        consecutive = 0
        while True:
            counter +=1
            if(np.random.rand()<p):
                consecutive +=1
            else:
                consecutive = 0
            if consecutive == 2:
                break
        trials.append(counter)
    return(np.mean(trials), np.var(trials))

In [18]:
expectation = 0
moment2 = 0
for i in range(1000):
    expectation += i*calculate_probability(i)
    moment2 += i*i*calculate_probability(i)

In [19]:
print(moment2-expectation**2)

22.00000000000002


In [20]:
simulate_experiment(1000000)

(5.996216, 21.962195681344006)